src/components/logic/relationship.py

```diff
@@ -0,0 +1,136 @@
+        """Extract the activity set from the log."""
+        activities = set()
+
+        for trace in self.log:
+            for activity in trace:
+                activities.add(self.activity_concept_name(activity))
+
+        print(activities)
+        return activities
+
+    def extended_trace(self, trace):
+        """Convert a trace to the extended trace."""
+        return [TRACE_START] + trace + [TRACE_END]
+
+    # Activity related functions
+    def activity_concept_name(self, activity) -> str:
+        """Extract the concept:name of an activity."""
+        return activity[__CONCEPT_NAME__]
+
+    def is_start(self, activity) -> bool:
+        """Determine whether an activity is the start activity."""
+        return activity == self.activity_concept_name(TRACE_START)
+
+    def is_end(self, activity) -> bool:
+        """Determine whether an activity is the end activity."""
+        return activity == self.activity_concept_name(TRACE_END)
+
+    def is_extension_activity(self, activity) -> bool:
+        """Determine whether an activity is one of the extension activities."""
+        return self.is_start(activity) or self.is_end(activity)
+
+    # Trace related functions
+    def is_empty(self, trace) -> bool:
+        """Return whether the trace is empty."""
+        return len(trace) == 0
+
+    def project_trace(self, trace, elements):
+        """Project the trace to a given set of activities."""
+        res = filter(
+            lambda ac: self.activity_concept_name(ac) in elements, trace)
+
+        return list(map(lambda ac: self.activity_concept_name(ac), res))
+
+    def first(self, trace):
+        """Return the first activity."""
+        return trace[0]
+
+    def last(self, trace):
+        """Return the last activity."""
+        return trace[-1]
+
+    def count(self, trace):
+        """Return the number of activity in that one trace."""
+        return len(trace)
+
+    def apply(self):
+        """Implement a relationship algorithm."""
+        results = []
+
+        # trace = [a, b, c]
+        # trace x trace = [(a, a), (a, b), ..., (c, a), (c, b), (c, c)]
+        source = itertools.product(self.activities, self.activities)
+
+        print(self.log)
+        for a1, a2 in source:
+            res = True
+            for trace in self.log:
+                res = res and self.apply_to_trace(trace, a1, a2)
+
+            if res:
+                results.append((a1, a2))
+
+        # Flatten out the list
+        # Each apply_to_trace returns a list of results.
+        #   results might look like -> [[(a, b), (b, c)], [(b, d)], [(c, a)]]
+        # Flattening this example leads to: [(a, b), (b, c), (b, d), (c, a)]
+        # flattenedResult = [val for traceRes in results for val in traceRes]
+
+        return results
+
+    def apply_to_trace(self, trace, a1, a2) -> bool:
+        """Apply the matching algorithm to each pair of activities."""
+        if self.activity_pair_matches(trace, a1, a2):
```

Comment:
Maybe I am missing something but doesnt activity_pair_matches only take 2 parameters in the current implementation?

Reply:
I forgot to change it in the abstract class. 😅 The implementation in the subclass takes 3 parameters (that is how it is suppossed to be).
I'll change that real quick! Thank you

Reply:
Ok great! I'll approve it once its changed.